### ISSUE TITLE
allow configuration of line length in pubspec.yaml

### ISSUE THREAD

**Issue by lukepighetti** [dart-lang/dart_style#918]
Hey @munificent is it possible to allow configuration of dartfmt line length in pubspec.yaml or some other common project configuration file? Of the last 5 flutter packages I've pulled down to work on 2 were using a 120 character line length which is becoming more and more common in Flutter projects.

I understand that we have access to IDE settings for line length but this is non-viable because it's IDE specific and not project-specific. Project specific would allow a package maintainer to make an increasingly common style choice and have all maintainers automatically use it without having to worry about IDE settings.

Any thoughts?

```yaml
name: foo
description: An example app.
version: 1.0.0

environment:
  sdk: ">=2.6.0 <3.0.0"

dartfmt:
  line_length: 120

```

**Comment by munificent** [comment]
Hi, Luke. I think you can probably guess what my thoughts are: https://github.com/dart-lang/dart_style/wiki/FAQ#why-cant-i-configure-it :)

If I were to add support for this, the very next day, all progress across the entire Dart ecosystem would screech to a halt as team members spent all day debating what line length should go into their package's config. I have seen no data to support the idea that programs vary such in their form such that it is measurably better to have different line lengths for different programs.

I took a corpus of the 994 most recent pub packages as of January. Of them, 763 contain ".dart" files with at least one line longer than 80 characters. When I exclude lines that start with "import", "export", or "part" that drops to 756. Skimming the result, most of long lines seem to be doc comments or long string literals, both of which aren't split by dartfmt.

If I exclude lines containing "//" or a quote character, then only 335 (34%) packages contain any files with a line over 80 characters.

I went ahead and looked at the individual lines of code in every file. Ignoring any trivial line that contains only whitespace or a punctuation character or two, I can bucket the remaining lines into "short" (<= 80 chars) and "long" (> 80).

212 (21%) packages have a ratio of 1% or less long lines to short. Only 62 (6%) packages have more than 5% of their non-trivial lines over 80 characters.

This says to me that most users are comfortable fitting within the current line length. And, even in packages that do go over 80 characters, it seems to rarely actually be helpful. Even packages that do have long lines, indicating that the user didn't format or formatted with a longer length have very few lines that actually exceed 80. That means the wider page size isn't buying them much except for wasted screen real estate on the majority of lines that don't benefit.

**Comment by lukepighetti** [comment]
I can see that you've spent a lot of time thinking about this. The problem with your stance is that projects already do make the decision whether to use the default line length or a custom one, the problem is that when they make that decision they aren't able to enforce it project wide. So not allowing line length settings in pubspec actually degrades the quality of dartfmt in practice compared to leaving it up to IDE config files.

**Comment by munificent** [comment]
> projects already do make the decision whether to use the default line length or a custom one

*Some* do, but most don't.

> the problem is that when they make that decision they aren't able to enforce it project wide.

Yes, that is a valid problem for those teams. But part of maintaining an open source project like dartfmt is deciding which use cases to support and which to not. Time is finite and time spent adding this feature is time not spent doing other work that benefits other users.

In this particular case, my belief is that the feature has net negative value across the entire ecosystem. I have yet to see any evidence that a line length longer than 80 columns actually improves productivity or readability to any measurable degree. It is easy to demonstrate that going wider than 80 columns makes it harder to do side-by-side diffs or two-panel editing on laptop screens (which is many of us right now and even in the best of times, many users cannot afford large displays). We can directly observe the the time spent arguing over style choices in code reviews and elsewhere when users are given the ability to make style choices.

I understand that this particular issue is very close to your heart. But my very clear experience over several years of maintaining dartfmt, personally interacting with hundreds of users, and seeing dartfmt enforced across millions and millions of lines of code is that the best experience for all users is to just let it do its job, move on with your life, and focus on what your code does and not how it looks.

**Comment by onyno** [comment]
In VSCode, I prefer a long line length (120) but I also want the structural formatting that comes with dartfmt, to get that I have the following set which passes in the command line -l option.

`"dart.analyzerAdditionalArgs": [
    "-l 120"
  ]`

In addition to the `"dart.lineLength": 120` and the `"[dart]": { "editor.rulers": [120]}`

**Comment by ajay-mm** [comment]
@munificent Making dartfmt use line length from pubspec does not in any way create any issues for those developers with 80 char width devices. It is an easier way of enforcing 80 character limit for projects that want to enforce it, instead of asking every developer to change their ide settings or add the command line argument while formatting.

You can keep your side by side diffs for projects you work on. But don't enforce your constraints on other people. @lukepighetti  isn't asking you to change the defaults. He is only asking for a very simple, straight and sensible way to do something which is already possible today but requires lot of manual efforts for everyone involved.

**Comment by lukepighetti** [comment]
As of today line length is user configurable but its fractured and unreliable because it can only be set in IDE configurations (like `.vscode/settings.json`) but it cannot be set in `pubspec.yaml`. If we could set it project-wide in `pubspec.yaml` then any user of any IDE would be able to minimize diffs when working on a project. And the 80 line limit is an arbitrary and unfair constraint on modern developers.

**Comment by israellias** [comment]
+1

**Comment by Antoniossss** [comment]
{quote}
AVOID lines longer than 80 characters

Readability studies show that long lines of text are harder to read because your eye has to travel farther when moving to the beginning of the next line. This is why newspapers and magazines use multiple columns of text.
{quote}

But programming is NOT a newspaper. If 80 chars line lenght is that great, why people keep asking to change that from pubspect (and in the meantime changing that per IDE instance)?
Studies done on (newspaper) readers are applied to programers. I personally sometimes feel that 160 is too small, but thats what I am using for most of my projects.

+1, allow to change that and lower the "readibility" score.

**Comment by crimsonvspurple** [comment]
Please allow line length setting to be set in pubspec as not having this causes problems for anyone no wanting to use default 80 chars.

With ultrawides being more common and descriptive class/variable names, 80 chars limit is too low. We use 160 for all our projects.

**Comment by flukejones** [comment]
This is becoming very frustrating to deal with. We no-longer use 4:3 800x600 displays, and have 16:9, 16:10, or even 21:9 displays at high res which can comfortably use higher that 80 char lines, in multiple views.

80 char line limits become *even more of a problem* when it unnecessarily breaks a line that is fine at say 100 char because the variable names are perhaps a little longer for descriptiveness (e.r, `status.isPermanentlyDenied`). Such instances reduce readability quite drastically.

80 char line limit enforcement across a *whole* ecosystem is frankly baffling. And archaic.

**Comment by flukejones** [comment]
It's as simple as this:

80 char limit:
```dart
        builder:
            (BuildContext context, Box<MdlMachine> machines, Widget? child) {
          return OrientationBuilder(
              builder: (BuildContext context, Orientation ori) {
            var nameInBody = false;
```       

100 char limit:
```dart
        builder: (BuildContext context, Box<MdlMachine> machines, Widget? child) {
          return OrientationBuilder(builder: (BuildContext context, Orientation ori) {
            var nameInBody = false;
```

I gain nothing from being forced in to 80 char limits here except for wasted lines. And the further you get in to flutter the worse it can get, particularly since widget properties can become verbose and are often required to be descriptive to convey their meanings and requirements easily.

**Comment by munificent** [comment]
> Please allow line length setting to be set in pubspec as not having this causes problems for anyone no wanting to use default 80 chars.

dartfmt does not have any awareness of pubspecs. Right now, it just formats each file individually completely context-free. This makes it easy and fast to run it in a variety of configurations. For example, many text editors pipe source to it through stdin. Changing dartfmt to scan parent directories for pubspecs would be a significant change in how it behaves.

> 80 char limit:
> 
> ```dart
>         builder:
>             (BuildContext context, Box<MdlMachine> machines, Widget? child) {
>           return OrientationBuilder(
>               builder: (BuildContext context, Orientation ori) {
>             var nameInBody = false;
> ```
> 
> 100 char limit:
> 
> ```dart
>         builder: (BuildContext context, Box<MdlMachine> machines, Widget? child) {
>           return OrientationBuilder(builder: (BuildContext context, Orientation ori) {
>             var nameInBody = false;
> ```

Following the [recommended style](https://dart.dev/guides/language/effective-dart/design#dont-annotate-inferred-parameter-types-on-function-expressions) would give you:

```dart
        builder: (context, machines, child) {
          return OrientationBuilder(builder: (context, ori) {
            var nameInBody = false;
```

**Comment by crimsonvspurple** [comment]
Default can be 80 or 60 or 100. That's is fine.

I believe the reason people are posting here about it is because you set longer line lengths. Commit push codes. Some day some other developer pulls, edits code, runs a format and things go back to default line length.

Basically, I (and probably many others) are asking for a way to have the line length property within the project itself so that regardless of who opens it wherever, code formatting stays the same. The default value or changing the default value in dart style guide is not related to this.

**Comment by ajay-mm** [comment]
For AndroidStudio atleast, we can now use .editorconfig which is supported out of the box.

We can set the line length like this

```editorconfig
# Editor configuration, see http://editorconfig.org
root = true

[*.dart]
max_line_length = 120
```

**Comment by lukepighetti** [comment]
Just had a situation where someone submitted a PR to my repo with their own line length. If line length were project specific and not left to IDE settings, I believe this could have been avoided

https://github.com/lukepighetti/salomon_bottom_bar/pull/15

**Comment by lukepighetti** [comment]
Also, had a chat with a Kotlin dev the other day who was lamenting that we couldn't have Dart on the same line length as our Kotlin and Swift codebases because there was no way to configure it at the project level

**Comment by lukepighetti** [comment]
Ran into this again today, this repo is using >80 char line length. I have no problems reading and manipulating the code. But because there is no pubspec.yaml setting for line length, my IDE is breaking formatting and increasing diffs unnecessarily. https://github.com/osaxma/dfs

**Comment by munificent** [comment]
Copying my comment from [here](https://github.com/dart-lang/pub-dev/issues/3956#issuecomment-1010553599):

Dart format does not have any notion of "project" or "package" and doesn't know where to look for any sort of global or package-level configuration data. It just formats each file path you give in a completely context-free manner.

This is important because:

* **It's faster.** The formatter doesn't look through parent directories for config files and determine which package a given file belongs to before it can determine how to format it.
* **It ensures you get consistent formatting when formatting files on disk or in your editor.** A lot of editors invoke dart format by passing a string of code to it on stdin and reading the result back from stdout. For those invocations, there is no file path that the formatter could use to find the config file. We could require IDEs to pass in the desired line length, but that's likely to be brittle.

We have millions and millons of lines of Dart code inside Google formatted to 80 columns. I can absolutely promise you that Googlers have come up with class names and method names as long as or longer than anything you see in the wild. This is nowhere near the top of the priority list for improving Googlers' lives and I find it very hard to believe that configurable line length will measurably improve the lives of any external users either.

I have never seen code that I felt would be easier to read with a longer line length. I've seen a lot of code that would be easier to read if the author rethought the verbosity of some of their identifiers.

**Comment by esDotDev** [comment]
> The problem with your stance is that projects already do make the decision whether to use the default line length or a custom one, the problem is that when they make that decision they aren't able to enforce it project wide. 

Exactly, I haven't been on a team that **hasn't** already had this debate. It usually lasts about 20 minutes, and we settle on 100 or 120. 

This debate is already being had, we just don't have a way to enforce it. Which is much more of a productivity blocker than the initial debate, as each time a new team member is onboarded to a project, they start screwing up the git diffs.

**Comment by munificent** [comment]
> This debate is already being had,

This is true *for some teams*, but not all. If we add support for configuring it the formatter, than we will expand the set of teams that end up having this debate. That's an anti-goal for the formatter. One of the key reasons this software exists *at all* is to reduce the set of things teams debate.

Imagine that someone passed a law—literally a criminal offense—if you formatted your code longer than 80 columns. You were forced to comply. Do you honestly believe that your engineering productivity or overall quality of life as a human on Earth would be measurably negatively impacted? If not... why not just stick with 80 and spend your limited time on this spinning planet on stuff that matters?

**Comment by lukepighetti** [comment]
My stance is that configurable line length is half implemented. We should either remove the ability to format at custom line lengths entirely, or complete the implementation and allow it to be set at the project level

**Comment by esDotDev** [comment]
> Do you honestly believe that your engineering productivity or overall quality of life as a human on Earth would be measurably negatively impacted? If not... why not just stick with 80 and spend your limited time on this spinning planet on stuff that matters?

Measurably is the key word here, but yes, my day to day enjoyment on my ultrawide monitor is negatively impacted substantially when reading and writing Flutter code that needs to break at 80 lines.

Not so much package code, because package code tends to be less indented overall. But Flutter day-to-day UI code? It is quite annoying yes.

Could I live with it? Of course, but I don't find that a very convincing argument. This is a discussion about quality of life discussion where small things matter.

> we will expand the set of teams that end up having this debate.

True, but you miss a key benefit in your overview hear. For the teams that do have this debate, you will enhance their productivity because they can properly implement the results of their debate and be done with it.

**Comment by munificent** [comment]
> We should either remove the ability to format at custom line lengths entirely, or complete the implementation and allow it to be set at the project level

Many many many command line tools accept options on the command line that aren't also available from an implicitly discovered configuration file, so I don't see accepting the line-length argument as being an incomplete implementation.

> For the teams that do have this debate, you will enhance their productivity because they can properly implement the results of their debate and be done with it.

At the expense of *lowering* the productivity of every user that switches between multiple codebases that would configure their line lengths differently. Every time you clicked "go to definition" on some imported library and ended up in a file whose line length doesn't fit in your IDE window, you're having a bad experience.

The charter of dart_style is to help the Dart *ecosystem*, not just to improve the lives of individual developers. It is a public park, not a lawnmower you can use for your lawn. All are welcome to visit its park and *because* it is a shared park used by many, it can be particularly well manicured. But visiting it means accepting that maybe the grass is not cut to the exact length you might personally prefer. Hopefully, being able to explore an entire park (the entire package ecosystem) more easily makes that trade-off worth it.

**Comment by lukepighetti** [comment]
> Many many many command line tools accept options on the command line that aren't also available from an implicitly discovered configuration file, so I don't see accepting the line-length argument as being an incomplete implementation.

Looking at the context of code formatters, dartfmt is the only one I'm aware of that has a line-length argument but no way to set it for the project outside of IDE config files. https://prettier.io/docs/en/options.html#print-width

**Comment by flukejones** [comment]
An advanced search of github for `"max_width =" language:TOML` reveals line widths from 80 up to 160, with the most common being 100, for the `rustfmt.toml` that rust uses. And somehow this absolutely does not affect the Rust ecosystem.

None of the arguments made to keep 80 column limit as a *hard* limit make sense and it just inconveniences the owners of the repos. The "But googlers don't need it" argument is not as good as you think - are you going to tell  us to write shorter function and variable names to compensate for the forced line length also?

It is odd that a language as configurable as Dart, with a ridiculous variety of styles to write in, such that packages like "pedantic" are a thing, is getting so hung up on a single thing - line length. Change any one of [these options](https://dart.dev/tools/linter-rules) but don't you dare go over 80 char wide for lines.

**Comment by esDotDev** [comment]
> At the expense of _lowering_ the productivity of every user that switches between multiple codebases that would configure their line lengths differently. Every time you clicked "go to definition" on some imported library and ended up in a file whose line length doesn't fit in your IDE window, you'd have a bad experience.

No, not every user. Just users on very small screens, like 16" laptops. I don't know why we would optimize for the hobbyist developer that does not have a 2nd monitor attached. And if we are, why stop there, if you set it to 40 you could even include phone users. It seems pretty arbitrary, especially in a world where programmers bill out at $5000/wk, but a larger monitor to make you more efficient, is $200.

> visiting it means accepting that maybe the grass is not cut to the exact length you might personally prefer. 

I take your point about consistency, but the analogy feels more like a park that is ostensibly build for adults, but all the rides are made for people 4ft tall :p 

Maybe I'm being to cynical, but it really does feel like the underlying cause here is most of the googles use laptops, so it's optimized for small screens, and all of this other stuff is just rationalizing reasons for that optimization.

**Comment by munificent** [comment]
> The "But googlers don't need it" argument is not as good as you think - are you going to tell us to write shorter function and variable names to compensate for the forced line length also?

What I am suggesting is that your code would be *more readable to all who read it* if you considered shorter function and variable names and hoisting more code out of deeply nested subexpressions into local variables or helper functions. Using a wider line length is treating the symptom.

I have seen a lot of code that is clearly improved by the author rethinking names and nesting structure when they find it line wraps too often. I've never seen code whose readability I felt was improved by making the lines wider. Just because you have a bigger screen, that doesn't mean you have bigger eyes or a bigger brain. Long identifiers and deep expressions still take more effort to read, regardless of how they line wrap.

If you prefer particularly long identifiers and deep expressions, that's your choice. It's your code. But I don't think it benefits the larger Dart ecosystem to tailor the tooling around that choice. And I've never been convinced that it's a good use of the time that Google pays me for to implement support for configuring that in dart_style when I could be spending that time on other things.

**Comment by rydmike** [comment]
A "fairly" known Dart repo that uses lines much longer than 80 chars and don't even use dartfmt is of course the Flutter repo itself.

Which is kind of ironic since dart and flutter packages on pub get a score penalty if dartfmt and max 80 char line length is not used.

I think both @lukepighetti and @esDotDev made excellent points on why being able to set the line length on a project level for dartfmt would be very useful.

I would definitely prefer to use eg 120 on a project level too, because we use the lint that the Flutter repo does, the one that requires you to specify all types. This extra, but in our opinion nice formal verbosity, definitely adds to line lengths and causes more not so terrific looking line breaks. It would be nicer at 120, but we have stuck to 80 due to the fact that it cannot be set on project level.

**Comment by esDotDev** [comment]
The truth is, leaving teams in a state of technical debt because they can not properly specifiy a team-wide setting like line-length, is an order of magnitude more disruptive than having a debate about line-length in the first place. 

And as luke mentioned, line length is one of a dozen different discussions you need to have when discussing line rules in dart for any project. It's awesome that we can define all of these options as our team desires, but then bizarre that we are blocked on setting something as intrinsic as preferred max line length.

If the argument is _truly_ about developer efficiency, there is no debate imo, this policy does far more harm than good.

**Comment by crimsonvspurple** [comment]
At this point it's clear whoever's in charge of this, isn't gonna budge. It's very likely the people from flutter had this debate with dart guys and since flutter doesn't use dartfmt; we can infer the outcome 😒

I'm glad I use Intellij where I can just commit settings file and have it format with 200 line length every time. I'm also glad I'm in position to enforce everyone in the team uses the same IDE.

If I ever push something in the "public park", people who don't like the size of the grass can stay out for all I care.

**Comment by munificent** [comment]
> It's very likely the people from flutter had this debate with dart guys and since flutter doesn't use dartfmt; we can infer the outcome 😒

The Flutter team has never used dartfmt. I don't recall any discussions with them about line length but as far as I know, it wasn't a major factor in them choosing not to use it. (The main reasons as I understand it are argument list formatting and indentation.)

**Comment by esDotDev** [comment]
An interesting field report, to show what _actually_ is the fallout from this decision across the garden, every single day.

Someone submits a PR:
```
[ Maintainer ] "what line length are you using in your editor? Unfortunately, the whole source got reformated so I don't see what was changed."

[ Submitter ] "Oops, sorry about that. Did not know it could become an issue. I did set it to 120 while trying to make use of my Ultrawide monitor. Will set it back to 80. I can re-submit if you want?"

2 hours later.

[ Submitter ] "Okay, this is not working. I'm going to have to submit a completely new one in order to fix the line length issue without having changes all over the files."
```

Just think about how much of everyones time is being wasted here in this simple interaction. The submitter submitted a doomed PR. The maintainer got an email, wasted his time looking at the src, had to respond, a conversation, ensured, the submitter spent a non-insignificant time trying to make it work, but in the end just doing a whole new PR was easier. 
https://github.com/escamoteur/get_it_mixin/pull/20#issuecomment-1030193899

Like at some point, pragmatism really needs to win out here over theory. Our projects are highly configurable already, and the setting is there, just let us set it so we can avoid this waste of valuable time.

**Comment by lukepighetti** [comment]
@munificent Should this issue be in different repo? I'm not privy to the toolchain graph. Which part of the toolchain would be most responsible for reading pubspec.yaml and dispatching formatting requests to dartfmt?

**Comment by munificent** [comment]
This is the right repo. If a tool would be reading the pubspec for formatting, it would be in dart_style.

**Comment by mx1up** [comment]
in our team we use 120, which is enforced throughout the build pipeline when merging. But every time build_runner builds, it reformats the generated files to 80. Having it stored somewhere like suggested here would allow build_runner to read this value

**Comment by NANASHI0X74** [comment]
I've started developing with flutter again after a long pause yesterday.
I previously wasn't aware that dart tooling had any defects of this caliber, but that must have been because essentially I was previously the only dev for the codebase.

> [ Maintainer ] "what line length are you using in your editor? 
> Unfortunately, the whole source got reformated so I don't see what was changed."
> 
> [ Submitter ] "Oops, sorry about that. Did not know it could become an issue.
>  I did set it to 120 while trying to make use of my Ultrawide monitor. 
> Will set it back to 80. I can re-submit if you want?"
> 
> 2 hours later.
> 
> [ Submitter ] "Okay, this is not working. I'm going to have to submit a completely 
> new one in order to fix the line length issue without having changes all over the files."

I can relate. In our codebase we have a readme file:
```
## Development Environment
- Increase maximum line length for Dart in Android Studio to 120
    - Editor > Code Style > Dart
```
...
I use emacs btw.

Hey @munificent.
>Yes, that is a valid problem for those teams. But part of maintaining an open source project like dartfmt is deciding which use cases to support and which to not. Time is finite and time spent adding this feature is time not spent doing other work that benefits other users.

Thanks for wasting my time.

**Comment by NANASHI0X74** [comment]
I understand that your time is also limited, but what I really can't stand is people who state that they know exactly what's the best decision for people who are not them. 

>In this particular case, my belief is that the feature has net negative value across the entire ecosystem. I have yet to see any evidence that a line length longer than 80 columns actually improves productivity or readability to any measurable degree. It is easy to demonstrate that going wider than 80 columns makes it harder to do side-by-side diffs or two-panel editing on laptop screens (which is many of us right now and even in the best of times, many users cannot afford large displays). We can directly observe the the time spent arguing over style choices in code reviews and elsewhere when users are given the ability to make style choices.

😡🤯🤬 
Cut that out. And be aware that you're not only seeing just the tip of the iceberg when you look at all pub packages but also a particular subset likely with the most extreme of selection biases. Production code looks a lot messier than open source libraries

**Comment by esDotDev** [comment]
`many users cannot afford large displays` - This is an odd statement that seems to favor hobbyist developers. 

In a professional capacity, purchasing a large display amounts to about 1/2 day of billable work. 

I'm unsure why we're letting such subjective rationale decide the day here.

**Comment by NANASHI0X74** [comment]
> `many users cannot afford large displays` - This is an odd statement that seems to favor hobbyist developers, or tinkerer coders.

as far as i see it's a waste of time even discussing this argument, because what's being asked is not to change the default, but to add the option to configure the dartfmt line length deterministically for every dev on a project via a config file checked in to the repo.
The experience for anyone who does not choose to use this option of course should not change.
I see the point the point somewhat
> `many users cannot afford large displays` - This is an odd statement that seems to favor hobbyist developers, or tinkerer coders.

> If I were to add support for this, the very next day, all progress across the entire Dart ecosystem would screech to a halt as team members spent all day debating what line length should go into their package's config

Despite this being a ridiculous hyperbole, I see the point somewhat that this might encourage bikeshedding discussion, but I think the amount of bikeshedding this will actually cause in addition to what's already happening and the time spent on that is miniscule compared to the time developers spend on coordinating and configuring their editors consistently because their tools are failing them. And this despite the fact there is a standard for this with editorconfig.
And I'm not saying that dartfmt should support all the editorconfig features, but when the formatter already has a command line argument for that, it is exceptionally stupid not to make it deterministically configurable via config options.
Every linter is also configurable because the needs of all developers and all projects **_are simply not the same_**

**Comment by NANASHI0X74** [comment]
And btw @munificent I read this: https://github.com/dart-lang/dart_style/wiki/FAQ#why-cant-i-configure-it
And I have to say that what it says is false- the way dartfmt currently works violates the first stated goal of the formatter because it hinders teams from consistently formatting their code because dartfmt is the standard way of formatting dart code. (on a related note, the fact that the FAQ is saying that dartfmt isn't right for every dart developer implies that dartfmt should not be part of the core dart tooling)

>This may be surprising, but the goal of dartfmt is not to automatically make your code look the way you like. It's to make everyone's Dart code look the same. The primary goal of dartfmt is to improve the quality of the Dart ecosystem.

I also see that point. But there are better ways to enforce this than to waste everyone's time configuring their shitty editors and third-party formatters. You already have the pub points system, simply use it and give points to packages that observe dart style conventions and defaults like the 80 character limit.

And that the configuration violates goal 3 and half of goal two is also bogus because as stated, the default should be the same and debates will either ensue only when there's a reason for it or even worse, there will be no debate, but frustration, wasted time and shitty third-party tools galore as is currently the status quo.

**Comment by devoncarew** [comment]
@NANASHI0X74 - several of your comments above violate the code of conduct for the community. We expect everybody to be respectful and courteous in their communications. Specifically:

- Respect people, their identities, their culture, and their work.
- Be kind. Be courteous. Be welcoming.
- Listen. Consider and acknowledge people’s points before responding.

https://dart.dev/code-of-conduct

**Comment by NANASHI0X74** [comment]
Thanks for the advice. I realize that my frustration has led me to violate point 2 in particular.
 I will restate the core point of what I said and do my best to be brief while observing the CoC.

https://github.com/dart-lang/dart_style/wiki/FAQ#why-cant-i-configure-it 
To me parts of the statements here are clearly erroneous, in particular the lack of configurability can and in this particular case definitiely does go against the first goal of having a formatter, because if some of the default parameters of the formatter are hard to change, people will turn to third-party formatters, in most cases their IDE, which may or may not be deterministically configurable via files in the repo. So that hinders people from 
>1. Produc[ing] consistently formatted code.

Tales of problems that have caused frustration and loss of valueable (life)time are abundant in this thread.

**Comment by NANASHI0X74** [comment]
To solve problems like this, there are certain standards, in this particular case chiefly editorconfig.org 
it is a tremendous pity and, I believe, a defect that dartmft doesn't support that standard at all.

**Comment by flukejones** [comment]
I truly do not understand why this is not a solved thing yet. It is enormously frustrating to begin contributing to another dart library only to discover that dart fmt does not format in the way the repo authors do, resulting in *huge diffs of unrelated changes*.

For the love of all developers out there, please enable a config option that people can set so that their chosen formatting is upheld.

**Comment by Daybayzayd** [comment]
He would lost less time developping this functionality that is, surely, very useful, than posting "argument" against that...

**Comment by GaithAST** [comment]
I can sum up this conversation with the following

- Everyone: can we have configurable line length in pubspec.yaml (harmless by the eyes of mostly everyone)

-- Dart Lang: No, that'll hurt the unification and harmony of dart packages

- But we already changed them, it's just an unnecessary pain were going through

-- It will hurt productivity

- Don't worry, it wouldn't harm it, we already agreed.

-- It would hurt the ecosystem

- Actually, it would benefit it as the contributer won't need to worry about line length

-- It would hurt people with small screens 

- But most of us have very good screens now

There is no sane reason we shouldn't be able to change it easily project-wide

**Comment by esDotDev** [comment]
The issue with PRs and bad diffs should be enough to close the conversation imo. 

If you're wasting package maintainers time, by not letting them set a line-length for their project, spread across the entire ecosystem, you're wasting an order of magnitude more time than any small discussion that might be had at the start of any given project.

**Comment by gslender** [comment]
Isn't the point of an open community is to listen to the voice of the community and adapt accordingly?

Still struggle to see how adding an option is a problem. Sure, if the request was to change the current default from 80 to 120, yeah that would be an issue, but making an option of a secondary place to determine dart-fmt defaults hardly seems a big deal - certainly not as big a deal as being made by the maintainers of Dart Lang. 

Smells of an academic argument where the theory might be right, but the practical real-world reality is way different.

Let's get real hey?

**Comment by dkbast** [comment]
I think there could be another way to resolving this discussion:

dartfmt is the underlying tool which is agnostic of its surroundings, that seems to be a good thing for performance reasons, lets keep it that way and maybe "expose" it again so it can be called directly.

But when invoking it using the CLI for flutter or dart (flutter format / dart format) we are in a situation which is context dependent and therefore I believe it would be a good thing to have all(?) the options which can go into dartfmt be configurable in the pubspec.yaml 

a workaround I'm currently using is having the format command be part of a makefile and call it instead of dart format, this way the configuration can be used consistently across different devices, but it's a pain when switching to projects which don't use this.

**Comment by matthewlloyd** [comment]
I'd like to add my vote for this. I'm deciding whether to use `dart format` for my existing codebase, and I would like to be able to choose a single line length that is consistent across multiple languages. I am with Linus on whether 80 is the right number.

`dart format`'s inability to customize line length at the project level is making me question whether it is a sufficiently mature tool for the job. The ability to do so appears to be something of an industry standard across multiple formatting tools used for many popular languages. For example:

* In Rust, the `rustfmt` tool allows `max_width` to be configured via `rustfmt.toml` at the project level: https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#max_width
* In Python, the very popular `black` tool allows `--line-length` to be configured via `pyproject.toml` at the project level: https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#configuration-via-a-file
* In Javascript, the `prettier` tool allows `--print-width` to be configured at the project level in a multitude of ways: https://prettier.io/docs/en/configuration.html

**Comment by chenasraf** [comment]
Would love to see this implemented. Since we can already define a line-length, it's not a decision about if that should be implemented or not. All we need is a permanent way to set the value without having to add CLI flags - which might be forgotten to be added on a per-developer basis. Sticking this in a `yml` file is a perfect way to do it... Don't really understand 3 years of conflict on this idea

**Comment by chenasraf** [comment]
If anyone is interested in a workaround, you can use a package like [script_runner](https://pub.dev/packages/script_runner) (disclaimer: this is my package).

1. Install globally:

    ```shell
    dart pub global activate script_runner
    ```

1. Add a script runner with the CLI args you want:

    ```yml
    # pubspec.yaml
    # ...
    script_runner:
      scripts:
        - format: dart format --line-length 100
    ```

2. And run it using `scr` instead of `dart`:

    ```shell
    $ scr format .
    ```

**Comment by dancamdev** [comment]
All I see here is needless resistance to a feature that would benefit many and detrimental to none.

You can still keep the default 80 line length limit, while allowing for others to change it as they please.

This is objectively an issue for many, project wide line length setting is IMO a basic feature that other ecosystems are solving efficiently - such as prettier using a prettierconfig.json.

Having something such as a 'dartfmt.yaml' would solve the issue and allow for configurations which I feel should be basic behavior for a formatter in 2023

**Comment by flukejones** [comment]
Even the Linux kernel has deprecated the 80 char limit https://www.phoronix.com/news/Linux-Kernel-Deprecates-80-Col

**Comment by caseycrogers** [comment]
I don't know if I have much to add other than +1 to allowing us to customize line length. For diffs, I already find split diff mode unreadable and never use it over unified. For writing code, I find 80 painfully short and really annoying to deal with, especially for Flutter code.

I get and respect the goals of consistency and short circuiting style debates on individual teams, but that's a tradeoff and I think line length is one of the places where the loss to consistency is justified by the gains to usability. All that said, I'd also be happy with deprecating the 80 line default and moving to a 100 line or 120 line default and still not allowing customizability-it's more the 80 that I have a bone to pick with than the inflexibility.

One thing I will say (which I saw elsewhere-I think from another Github issue) is that newspaper readability studies fundamentally do not translate to code readability. Namely, in a newspaper, the user has to scan the entire line width to read the line. In programming (ESPECIALLY in heavily nested Flutter), the start of the line is often a ton of whitespace for indentation that the reader need not scan to parse.

**Comment by gslender** [comment]
Yep, as maintainers of Dart, it feels like they don't do any Flutter development and have no clue what the big deal is having to make sense of totally unreadable 80 line length code.

Where else is Dart so heavily used more than in Flutter? Who's the folks that are benefiting from limiting the line length to 80?

It's a shame Google doesn't take its projects seriously as they'd be a greater force in the market place if everything wasn't addressed in academic ways.

**Comment by esDotDev** [comment]
What I find odd about this decision is that the rationale of saving developers time, is so obviously not being served by this decision. 

In an attempt to avoid any discussion among developers at the beginning of a project, this instead: 
* waste time by creating PRs that can not properly be diffed, this significantly slows down contributions to open src projects as the feedback loop between submitter and maintainer is often very slow. 
* waste time by requiring training for every developer on the project to set line-lengths in their IDE (android studio)
* waste time by forcing developers to change their IDE settings between different projects (android studio)
* we still have the discussion about line length anyways because everyone agrees 80 is ridiculous on a large monitor

**Comment by modulovalue** [comment]
**I would highly recommend to anybody having issues with the 80 char limit to open a separate issue with example code and a link to this issue for reference.** Remember that increasing the max line width might not be the best solution. The problem could be somewhere else. In other words, consider highlighting a problem, and not necessarily a solution.

---

> it feels like they don't do any Flutter development and have no clue what the big deal is having to make sense of totally unreadable 80 line length code.

Are you using optional trailing commas? In my experience they solve most of the issues that people have with the 80 char limit. Of course, not all of them, but even if the limit was 120 or 1200, increasing the max line width would not end up solving all of them and introduce new ones.

---

Thank you Bob for not giving in. I agree with you, I think that changing the default or making it more easily configurable would, in general, not make dart developers more productive.

**Comment by caseycrogers** [comment]
I figured sharing hard examples would probably be helpful here. As an experiment, I jumped into whatever part of my project I was last editing, took a screenshot, and then evaluated what places I feel 120 would improve readability:
Before:
<img width="500" alt="Screenshot 2023-06-08 151018" src="https://github.com/dart-lang/dart_style/assets/10675231/f019b4dd-bcf2-42de-9c99-665759c238a7">
After:
<img width="500" alt="image" src="https://github.com/dart-lang/dart_style/assets/10675231/e8db0e75-4e1a-41cc-af69-1dcb520b0aa8">

The three problems in this random snippet fit into two categories:
1. A function signature that is just a bit too long, so I have to add an optional trailing comma and split each argument onto it's own line. If the function had more than just two arguments, I'd find the multiline layout more readable, but not here. This one I must say is pretty 50/50. The line does get a bit overwhelming in the after picture.
2. Variable assignment that gets bumped onto the next line. I think this really hurts readability because the connection between the expression on the right and the variable on the left is a lot less visually apparent while skimming.

In general, each line split up by the char limit adds vertical space to my file which makes it a lot harder to navigate. I have empty space to the right so I'd love to use that empty space to reduce the amount of scrolling required and to get more code visible in the viewport at once. And, this screenshot was already taken with my standard layout that has a lot of other UI elements competing for space. Listed from left to right:
1. Project structure navigator
2. One sub-panel with files and tabs for editing (this is the screenshot)
3. A second sub-panel
4. The emulator
So if I'm already liberally using my horizontal space for other things and I still have room to fit 120 chars, it really makes me wonder how common it is that someone can only fit 80 chars?

I encourage others to share their examples (either for or against a 120 char option) so that we can get concrete examples to work with here.

**Comment by munificent** [comment]
> Yep, as maintainers of Dart, it feels like they don't do any Flutter development

I am a [readability mentor](https://www.pullrequest.com/blog/google-code-review-readability-certification/#:~:text=Readability%20is%20an%20internal%20certification,with%20readability%20in%20that%20language.) for Dart in Google. I have reviewed hundreds of thousands of lines of code in thousands of changes written by hundreds of engineers across dozens of applications. An increasing fraction of that code is in Flutter applications.

I review Flutter code for its readability almost every work day. For several years, I was the most prolific readability reviewer, which means I probably looked at more Dart code than any other Google employee.

In addition, as part of my job on the language team, I cultivate a corpus of tens of millions of lines of open source Dart code that I constantly analyze and skim through as part of the process of evaluating potential language changes. I have looked at a *lot* of Dart code written by many people.

> I would highly recommend to anybody having issues with the 80 char limit to open a separate issue with example code and a link to this issue for reference.

Yes, concrete examples of code that could clearly only be improved by increasing the line length would be helpful.

In almost all of the Dart code I've reviewed where lines are heavily wrapped, what I usually see is long redundant identifiers and overly complex expressions. One reason I prefer 80 columns as a default line length is that it incentivizes authors to *think about making their code actually simpler and clearer*.

**Comment by caseycrogers** [comment]
In this vein, looking at my above comment, the biggest culprit is definitely `DribbleDatabase.intance`.

This should be assigned to a local variable at the top which will make the following lines shorter and more readable and mitigate the line length problems I'm having.

**Comment by esDotDev** [comment]
> One reason I prefer 80 columns as a default line length is that it incentivizes authors to _think about making their code actually simpler and clearer_.

120 forces this same thought process as well, so this is not really a compelling argument. So does 60, or 40 etc. The question is what is reasonable and what becomes onerous.

I still think this comes back to work environment, do you use multiple 19 or 21" monitors as most production developers do? Or are you judging what is reasonable on a small 15" laptop?

On large monitors, the 80 char limit comes off as extreme, on a small laptop it seems reasonable. In this case this just becomes an argument for supporting easy configuration across a project, that way a team can make a decision based on their own work environments, and we're not trying to find some perfect number for everyone (which doesn't exist imo).

**Comment by rydmike** [comment]
The discussion of what is more or less readable when it comes to line length, while academically very interesting, is irrelevant to this feature request.

It does not matter what I, Bob, Modestas, Luke, Casey or Shawn argues is better, worse, more or less readable. 

What matters is what a software team has **agreed to use** and **enabling** them to **automate** and **enforce** that **agreed** line length on a project level.

The issues caused by not being able to enforce agreed line length of `dart format` on a project level, has been been exhaustively presented already.

Given a situation where a team has chosen to use another line length setting for `dart format` than 80 chars, the current feature gap causes unnecessary complications and frustrations for such teams and projects. It seems to me that finding a way to accommodate their preferences, would be in general interest of improving **Dart** tooling, be in-line with being open and inclusive, plus it would not take away anything from anybody.

This feature request does not propose to change the `dart format` default line length from 80 chars, nor does it propose any changes to the way `dart format` actually formats **Dart** code. It only requests that teams that have agreed to use another line length, should be able to do so **on a project level**.

----

What is more and what is less readable is of course academically an interesting topic. We could for example examine and discuss the code style of the **Flutter** project.

It does not use `dart format` at all, because the project has decided to format and break lines differently than what `dart format` does.

Flutter repo also uses extremely long lines, as and when it deems it appropriate to do so. I think this example of a **263 chars** long line, is probably one of the longer examples:

![Screenshot 2023-05-13 at 20 45 31](https://github.com/dart-lang/dart_style/assets/39990307/b2d135df-59f9-4fe8-9220-9275d4a87f3e)

We could argue if this is readable or not, but it is what the **Flutter** project has decided to use. 

If you try to submit a PR that uses `dart format` to the **Flutter** repo, it will be rejected, and you have to redo your PR following the Flutter repo formatting rules. This of course raises the bar to contribute to the project, since formatting submitted code cannot be done automatically at all. Which of course speaks to why `dart format` is typically welcomed by most projects and seen as a great tool, as it avoids such complications.

Having the capability to set line length setting on project level, would only strengthen `dart format`. Who knows, maybe one day even **Flutter** could use it.

**Comment by lukepighetti** [comment]
Just wanted to say that I have no interest in forcing other people out of 80 char. To be perfectly honest I would probably keep using 80 char. That said, there are projects using 120 char and even 140 char.

At work, we would use it to coordinate line length across Kotlin, Swift, Javascript, Go, Python which are all set to 120 characters. Dart is the only one we cannot configure at the project level. We can only configure it at the IDE level, and we (as a rule) do not force developers to use specific IDE settings. We only enforce/expect project settings. And they are not available to us.

So this issue is NOT about if 80 characters is good or bad.

This issue is about making sure we can configure --line-length at the project level. Whether that's in `pubspec.yaml`, `analysis-options.yaml`, or some novel `dartfmt.yaml`, is not clear to me.

@munificent, it would be helpful if we could get some clarity on if you would accept a PR to "close the gap" on this feature. And if the answer is yes, if you have any opinions on where this setting could live?

**Comment by gslender** [comment]
>So this issue is NOT about if 80 characters is good or bad.
>This issue is about making sure we can configure --line-length at the project level. Whether that's in pubspec.yaml, analysis-options.yaml, or some novel dartfmt.yaml, is not clear to me..

This is the point.

Hopefully common sense prevails and a PR is seriously considered along with the academic argument over whether 80 char line length could or should be used is left in the past where it belongs.

**Comment by esDotDev** [comment]
It's interesting to look at other languages to show how arbitrary this
limit is.

Default line lengths:
SwiftLint - 120
C# w/ Resharper - 120
Kotlin - 100
Go - 100
Python w/ BlackFmt - 88
Dart - 80

Of these, Dart is the shortest, and one of the only ones that does not
allow you to set it on a project level. One thing is clear from these
numbers: there is no perfect value.

Allowing you to change it, but then not define it at the project level
makes a poor middle ground. It opens the door to custom line lengths,
ensuring many(most?) teams will decide to change it, but then does not give
us the tooling to properly do so, which forces a lot of wasted time.

If the goal is to not waste time, then going fully in one direction or the
other would be more effective. Since the cat is already out of the bag with
regards to custom line lengths, it seems the only logical option is to
properly allow it to be set at the project level.

**Comment by flukejones** [comment]
Everybody commenting here wants the ability to configure the line lengths, regardless of their view on what is suitable.

But instead we are getting pigeon-holed in to an arbitrary line length that is essentially one persons opinion (and then they go on to argue by authority - we don't care, we just want the damned thing to be configurable by *us*).

**Comment by modulovalue** [comment]
Hi @caseyrogers,

Thank you for providing an example. I took a look and here are some thoughts on how I would deal with it.

I agree that it's not ideal that the formatter puts the await expression on the next line.

I would do the following to mitigate that:

- I personally would remove the type and let type inference handle it.
- You seem to be using singletons to access a shared global instance of your database. I think that singletons are great for quick prototypes, but they are known to make code hard to test and hard to maintain because they introduce a dependency on global state. My suggestion would be to inject a db instance into the class via its constructor and to use that instead.

That way, your code should start to fit into 80 chars.

For example:

```
final ... db;

...({
  required this.db,
});

...
  // 59 chars
  final cachedGuesses = await db.getStoredShotClockGuesses(
...
```

In short, I would suggest to avoid singletons and to consider making more use of type inference. Increasing the max line width limit would of course also help with your example, but I think that the alternative solution (i.e. to not use singletons) is much better.

**Comment by gslender** [comment]
This is getting to the point where it might be better for us to all escalate to someone senior in Google who cares more about the community - I can't believe line length is being argued as a mechanism to improve quality of the code - I must of missed that lecture at university!?

**Comment by gslender** [comment]
@kevmoo are able to look into this and provide some guidance around evolving Dart in a way such that the community appreciates the direction?

**Comment by Daybayzayd** [comment]
With line break :
<img width="501" alt="00_80_line_length" src="https://github.com/dart-lang/dart_style/assets/55182867/69eb3f06-d8c3-497b-a93a-001f85bfd327">

Without line break :
<img width="541" alt="01_80_line_length" src="https://github.com/dart-lang/dart_style/assets/55182867/5df01b70-324e-4d87-864a-3769806b9381">

It's more readable in the second screen, and the number of line is not increasing.

**Comment by modulovalue** [comment]
@Daybayzayd It looks like everything is working as intended with respect to the observations discussed in this issue. (Or maybe I'm not understanding what you were trying to bring across)

You can remove the trailing comma if you want to tell the formatter not to format Text widgets in a way that makes them span multiple lines.

```
//                                    ,- if you remove this, the formatter will not put your Text Widgets on multiple lines
//                                    v
Text("some text", style: _contentStyle,)
// Like this:
Text("some text", style: _contentStyle)
```

**Comment by Daybayzayd** [comment]
> @Daybayzayd It looks like everything is working as intended with respect to the observations discussed in this issue. (Or maybe I'm not understanding what you were trying to bring across)
> 
> You can remove the trailing comma if you want to tell the formatter not to format Text widgets in a way that makes them span multiple lines.
> 
> ```
> //                                    ,- if you remove this, the formatter will not put your Text Widgets on multiple lines
> //                                    v
> Text("some text", style: _contentStyle,)
> // Like this:
> Text("some text", style: _contentStyle)
> ```

Thank you so much. I didn't know this rule. It works. I edited my post (and not deleted because i think if one like me is concerned by this problem, he can maybe see the answer :D )

**Comment by TimWhiting** [comment]
I don't think it is a problem for the build.yaml file to also allow a global dart-line-length option. Of course the generators still need to respect this option, but I think the community would be willing to update the generators to use this option when formatting the generated files.

**Comment by caseycrogers** [comment]
> Hi @CaseyRogers,
> ...
> In short, I would suggest to avoid singletons and to consider making more use of type inference. Increasing the max line width limit would of course also help with your example, but I think that the alternative solution (i.e. to not use singletons) is much better.

Thanks for taking the time to respond with your thoughts!
I won't get too into it so as not to move too far away from the topic at hand, but I like singletons and have a code template that generates them with an `@visibleForTesting inject` method. I also like being very explicit about types-I find the increase in verbosity is made up for by how much easier it is to keep track of what a variable is (there's even a linter rule for requiring explicit types).
WRT line lengths, I assigned my singleton to a local `db` variable and it solved one of the two awkward line wraps-a definite improvement!

Unrelated to the above:
this conversation is getting pretty heated, lets try to keep the conversation cool (like Fonzi). Heated conversation just makes everyone less likely to engage and we're much less likely to get configurable line length if we argue for it in a heated manner!

**Comment by modulovalue** [comment]
It looks to me like a big downside of Flutter using dart_style would be that every time Bob decided to approve a change, the world in Flutter-land would be forced to stand still for a moment as all contributors would have to update their PRs and learn about the new formatter changes. Whereas on pub.dev, there's at worst a pub-point penalty for not using dart_style, and everybody else is free to choose how much they want to depend on dart_style.

In other words, it seems to me that making the max line width configurable (via e.g. the pubspec file) would not help in convincing Flutter to adopt dart_style.

---

> WRT line lengths, I assigned my singleton to a local db variable and it solved one of the awkward line wraps-a definite improvement regardless of the max allowable line length!

I'm glad that that suggestion was helpful. I agree that singletons can be useful and I agree that having explicit type annotations everywhere can be helpful (especially for understanding code on GitHub or in presentations where the dart analyzer is not available).

---

I also strongly agree with Jake. There are other code generators that are not `build` and like to use `dart_style`. Forcing them all to to have some custom configuration format for line widths and keeping them all in sync and teaching all users that they should keep them in sync sounds like a lot to ask for. And a new standard in the spirit of [NO_COLOR](https://no-color.org) (e.g. a DART_FORMAT=100 environment variable) could be problematic when working with other people (i.e. environment variables are generally not checked into source control systems so it would introduce sync issues between members of a team).

**Comment by natebosch** [comment]
> Allowing you to change it, but then not define it at the project level makes a poor middle ground. It opens the door to custom line lengths, ensuring many(most?) teams will decide to change it, but then does not give us the tooling to properly do so, which forces a lot of wasted time.

_Allowing_ a line length CLI flag does not "ensure" nor "force" it's use. If there are more ways that we can discourage use of this flag and save teams from wasting time we can certainly explore them. Removing the option entirely is not likely on the table because it's useful for tests, and it would be a significant amount of work for us to remove it and churn existing usage.

> an arbitrary line length that is essentially one persons opinion (and then they go on to argue by authority

The line length reflects the opinion in practice of early Dart developers. It was not chosen by one person, and it is not solely by one person's authority that it isn't changing.

> I think that pretty much covers it?

Other than hand-waving away the significant jump in complexity from not-configurable or aware of project layout, to configurable and knows how to navigate a project layout to find the config file, this is also missing consideration for all the ways that `dart_style` is used to format code outside of the CLI tool.

> I don't think it is a problem for the build.yaml file to also allow a global dart-line-length option.

When this cascades out into requests across the ecosystem for any use of `dart_style` to support per-project configuration, it amplifies the cost.

> Having something such as a 'dartfmt.yaml' would solve the issue and allow for configurations which I feel should be basic behavior for a formatter in 2023

The inevitable flood of requests we'd get for additional options in the file (which we'd have to turn down because we have no plans to increase complexity and write a configurable formatter) is another good reason not to introduce an entire configuration mechanism with a single configurable option.

**Comment by esDotDev** [comment]
> _Allowing_ a line length CLI flag does not "ensure" nor "force" it's use. 

If you have a default that is smaller than all of your peers, and expose the setting, of course you are ensuring that it will be used by many developers. I don't see anything controversial in that statement. 

Agree that removing it is not reasonably on the table, in which case it seems the only reasonable path forward is to finish what was started, and make the line-length properly configurable by project. This middle ground is the worst spot to be if you primary concern is saving developers time, which it seems is the core argument from Bob on this.

If we analyze this purely from a "dont waste developers time" perspective, the two best options are
A) Don't expose the setting at all
B) Expose the setting and allow it to be configured at the project level

Halfway between the two is the worst spot to be in terms of dev hours wasted. I would personally still prefer halfway over nothing at all, but I'm just pointing out the flaw in the analysis here. Once you have exposed it, then it only makes pragmatic sense that it should be configurable at the project level, the alternative is too messy.

**Comment by TimWhiting** [comment]
>> I don't think it is a problem for the build.yaml file to also allow a global dart-line-length option.

> When this cascades out into requests across the ecosystem for any use of dart_style to support per-project configuration, it amplifies the cost.

The point is that it isn't a cost the dart team has to worry about. There are developers out there who care about this, and it actually isn't that hard to create a PR that passes the right options into dart_style.

Is the only option then for people who care about this to create PRs for all of the code generators out there (possibly having to fork them), and fork (or wrap) a version of the dart format / flutter format cli to take into account a configuration file, and add support in a separate vscode extension (or maybe even in dart-code) to use a different formatter for dart files? I hate to suggest it, because I feel like the community should feel like their opinions are heard and valid. But it does seem like a viable alternative if this issue continues the way it has been. 

As long as there exists the option in dart_style to configure it, there is a way to make this nicer for developers. I just hope that a 1st party way to configure it seems more reasonable than having forks/wrappers of all these projects. I for one am tired of this issue just sitting here, and plan on contributing to the generators I use to allow this - side note, I've already done this for one code generator awhile ago (using an option in build.yaml), so I know the cost, and it is not much. The cost of having to turn on and off my user settings in vscode depending on what project I'm trying to contribute to (and revert 100s of changes that are just whitespace because I accidentally had a formatter turned on) is far more than updating a few code generators IMO. 

I know this ignores the workarounds I mentioned for non-code generator tools. However, I have been playing around with LSPs and vscode editor apis, and I really don't think the wrappers around the other tools would be too hard to create either. In fact I'd probably just contribute the changes to fvm.

> The inevitable flood of requests we'd get for additional options in the file (which we'd have to turn down because we have no plans to increase complexity and write a configurable formatter) is another good reason not to introduce an entire configuration mechanism with a single configurable option.

Yeah we get the message loud and clear. dart_style will never support more configurable options. If developers want more configurability they will need to create a fork of dart_style. 

If it is really a problem of having a file that exists with only one option available, why not put it in the .dart_tool hidden folder and just provide a cli option `dart format config --line-length=*`, which edits that single option. This way it is obvious that there is no 'configuration file', there is just a command line configuration which just supports one option. The config file is completely hidden from the user. As a side note, in order to make sure users / contributors have the same configuration, we might have to add a negation !.dart_tool/format.yaml to the gitignore, so that that single file is committed to the project. I see this as one viable way forward.

**Comment by lukepighetti** [comment]
@munificent @natebosch @jakemac53 I'm having a hard time understanding if the team would accept a PR that fit [my proposal above](https://github.com/dart-lang/dart_style/issues/918#issuecomment-1583706948), along with addressing the concerns from [Jake offered here](https://github.com/dart-lang/dart_style/issues/918#issuecomment-1584694057). It would be a shame for me to make this contribution and find out all along that it was DOA. Can someone give me clarification? cc @kevmoo

**Comment by larsbs** [comment]
@munificent I definitely don't have the answer to all those questions since I'm just a mere observer, but I have a couple questions more:

- Haven't most of those questions been successfully answered by the prettier team (judging by the amount of developers that are pretty happy with it)?
- Do you think they're facing different problems (and hence, need different solutions) than the ones we're facing in Dart?

**Comment by TimWhiting** [comment]
> This isn't a single-PR-scale change. Implementing this only in dart_style would make the command line `dart format` behave differently from the IDEs almost everyone is using. Most IDE integrations format files that are in memory, not on disk. They use the formatter's library API to do that, and in that mode, the formatter should probably not be wandering around the user's filesystem trying to find a config file. _The config itself_ may have an unsaved change to the line length which should be honored when formatting happens in an IDE.

Acknowledged, it would be best if these changes all happened at the same time, so that inconsistencies don't happen. I see a minimum requirement of dart_style, flutter / dart clis, vscode, IntelliJ, package:build, and package:source_gen. I acknowledge that all of these teams might not have the bandwidth, but I'm sure the community that cares about it has the motivation and time.

> In fact, the dart_style library _can't_ look at the file system without it being a major breaking API change. Currently, the dart_style library doesn't depend on dart:io, which allows it to be used on the web, Flutter, or other platforms or contexts where dart:io isn't available.

This points to a design requirement that it be implemented in the cli tools, not in the format code itself. There is already some io specific tools already in the dart_style package: [here](https://github.com/dart-lang/dart_style/blob/main/lib/src/io.dart). Seems like a good place to put the logic, and since it would be in the package, that logic would be reusable by dart cli, flutter cli, package:build, and package:source_gen. 
 
> To add support for this, we would have to coordinate the design with at least the IntelliJ and VSCode plug-in folks. They would need to add support for finding this config file through their own file system abstraction layer and sending the correct line length for each file when using the formatter's library API. I don't know if they have the bandwidth for that. Other editors with less active support may end up broken or inconsistent until/if they are updated too.

Agreed @DanTup, looping you in here. Would you have the bandwidth to review a PR if someone were to implement this.
As far as other editors go, if we can keep the logic small i.e. one config file, simple one line configuration which can be imported from dart_style/io.dart (this would need exporting a small interface) I feel that the community could work on support for editors that don't have official flutter support.

> There are design questions around the configuration itself:
> 
> * Does the configuration go in the pubspec? build.yaml? analysis_options.yaml? Somewhere else? What is its schema? How are configuration errors handled?

Using the pubspec.yaml is in the title of this issue. I know pubspec has grown beyond it's original design, but limiting it would break many plugins that I know of, so although it seems odd, I see no technical reason for not putting it there, though it would not be my first choice. build.yaml seems odd because it is typically only used when using code generators, and even then rarely, and with macros, more people might not end up using generators anymore. Generator specific line lengths could still be supported, in which case I would put those generator specific overrides in build.yaml, and the support in each of the generators themselves and only if they wanted. analysis_options is definitely related due to the linter warning about line length, it makes sense to have the configuration option next to the configuration enabling such a lint. As such I would lean towards an analysis_option.yaml configuration. As far as the format for the option I see no reason not to just keep it simple:

line_length: number

> * What happens if there are configs at multiple levels of the directory hierarchy? Does the one nearest to each file being formatted win? Or does the config in the same directory where the formatter is invoked win?

With the reasoning above about it being related to the linter configuration in analysis_options.yaml I would use the same analysis_option resolution.

> * How do code generators that use the formatter to format their results know what line length configuration to use? Who will own updating those generators to determine and pass in the right line length?

Ideally generators do not need to worry about it unless they provide their own custom builder, or an override for the default line length. I would personally add the logic to just package:build, and package:source_gen as a default for any generated file ending in .dart. (Maybe add an option to opt out). 
For example source_gen already supports a formatOutput option for builders [here](https://github.com/dart-lang/source_gen/blob/70e46c0079263192b7a914ac651f25eff4fa7680/source_gen/lib/src/builder.dart#LL49C35-L49C47). This defaults to using dart style [here](https://github.com/dart-lang/source_gen/blob/70e46c0079263192b7a914ac651f25eff4fa7680/source_gen/lib/src/builder.dart#L397). If dart style provides an io specific export that allows you to resolve formatter options for particular file paths, this could be updated to apply those options.

> * What happens when file IO or permission errors occur when looking up the directory hierarchy to find the config file? (These do indeed happen.) What if those errors are intermittent?
>   (Note that it's not enough for the implementation to handle this. The PR would have to _include tests_ of it. Writing file IO tests is time-consuming, annoying, and really hard to get working across all platforms. But if we don't have those tests, then it's too easy to regress that behavior.)

Agreed. File tests are buggy sometimes, but definitely necessary in this case. As for how to gracefully handle an error I would say emit a warning, and do not format. At worst, some updated code is not formatted (sometimes happens with VSCode anyways if the analysis server times out on format requests). However if it were to default to something that could change a lot of code that shouldn't be changed.

> * When looking for a config file, does it stop walking the directory hierarchy at some point and give up? (For example, Black stops when it sees a `.git` or `.hg` directory.)

Again, use analysis_option resolution. I think it is reasonable to stop early at a `.git` directory.

> * What happens if the user passes a `--line-length` argument _and_ there's a config file? Does that command line option specify a _default_ or an _override_?

Emit a warning, user should update the file so that the rest of their team can format the same as them. If they only specify a single file and not a directory, I would say treat as an override, but still emit a warning saying that the next time the file is formatted it will be changed unless you update the format config file, or manually override again.

> * What happens when users want to configure the line length differently for different parts of the same package? Is there a mechanism for this? Do we end up also needing per-file configuration?

The analysis_option.yaml file allows for ignoring / excluding files, and can be nested under directories. Use analysis_option scoping mechanism. Maybe formatting should have a separate exclude / include section - or most likely just an include option that overrides any excludes. Typically users exclude generated files, but I can see them wanting generated files to be formatted for readability of those files.

> * If we find a config file that _doesn't_ configure the line length, should we keep walking up the directory hierarchy looking for a surrounding one that does? Is the absence of a line length configuration treated as "configure this directory's files to the default" or "use any surrounding configuration"?

Yes, up to the package root. I'm assuming this is how analysis_option.yaml file works.

> * Assuming this goes in the pubspec, is there a way to configure the line length for a directory _without_ creating a pubspec? Should there be?

I'd prefer analysis_option.yaml due to how it answers all of the previous questions well, and addresses the scoping / including file mechanism.

> I'm sure all of these are tractable, but are decisions that would have to be made. I'm sure there are other design questions I'm not thinking of.

Happy to try answering any new questions

As far as where to put the bulk of the resolution logic: 

Given my leanings towards analysis_options.yaml, it would make sense for the analysis_server to expose a way to query options for line-lengths for specific files, which would simplify the editor plugin logic for those using the analysis_server, especially if it uses analysis_option.yaml based scoping. And dart_style already has a dependency on analyzer, which seems to fit perfectly. So the ultimate logic for resolution would be added to analyzer, with a small interface reexported from dart_style potentially (though package:build already depends on analyzer if I'm not mistaken), so that might not be strictly necessary.

Also, we could lean on the file based tests for analysis_options.yaml, with some additional tests for formatting specific configuration.

> If this was something I could just hack together in a couple of days and be done with it, I would have done it by now. It is actually a hard problem.
> It's not clear to me that solving it is worth the engineering effort, though I certainly see that some feel very strongly that it is.

Thanks for recognizing that some of us feel strongly that it is worth the effort. Let's say a group from the community had open PRs for 
- dart-lang/dart_style
- dart-lang/build
- dart-lang/source_gen
- dart-lang/sdk
- flutter/flutter
- dart-code/dart-code (vscode)
- IntelliJ -- wherever the plugin is

would that be acceptable? (obviously pending buy-in and approval from the vscode / intellij plugin authors as well).

I recognize that the proposal of putting it in package:analyzer, means that ultimately it becomes a dart team maintenance burden. I hate to do that to the dart team, when they have so many other priorities, and the language to improve. My hope is that formatting specific issues and maintenance could be annotated with a 'community:help' label, so that we know what issues the community can help with to give the dart team more time to work on nice language features and improvements to the compiler.

**Comment by DanTup** [comment]
> Agreed @DanTup, looping you in here. Would you have the bandwidth to review a PR if someone were to implement this for Dart-Code?

The formatting in Dart-Code is handled by the analysis server, so any work to support this would need to be done there (https://github.com/dart-lang/sdk/blob/main/pkg/analysis_server/lib/src/lsp/handlers/handler_formatting.dart). The server already supports setting the line-length, but it's currently LSP-specific way so running "dart format" won't read it.

Btw, there's a somewhat-related issue at https://github.com/Dart-Code/Dart-Code/issues/4482 about a "do not format" setting that Dart-Code used to have but doesn't work with LSP. The request is to be able to exclude some set of files from formatting and I suggested a `formatter` section in `analysis_options.yaml` (ofc Bob's caveats/concerns about reading files above apply here too) - if these things were to be implemented, it would make sense to try and align their config.

**Comment by TimWhiting** [comment]
@DanTup 
So a formatter section of the analysis_options.yaml, with a filtering excluding / include list, would work for that issue as well. 
```yaml
formatter:
  line-length: 120
  excludes: **/*.g.dart # Or whatever
```

As far as I know the analysis_server doesn't have qualms about reading files right (from the resource provider of course)?

> but it's currently LSP-specific way so running "dart format" won't read it.

As far as I understand it, this is the main issue. We need a package that both dart format / analysis_server can depend on for interpreting the options of an analysis_options.yaml file. They both depend on analyzer. I think this is an ideal location to put it.

**Comment by DanTup** [comment]
> As far as I know the analysis_server doesn't have qualms about reading files right

Correct, the server knows where pubspec.yaml/analysis_options.yaml are and can easily read them.

I'm not familiar with how the "dart format" command calls the dart_style API, but I wonder if the "dart format" command (which presumably lives in dartdev) would be a reasonable place to handle any config on disk, then provide arguments to the dart_style API?

**Comment by Nikzed** [comment]
I feel like it should be implemented as project property for sure, kinda hard to explain every new person to set line length to 120 each time, while doing code review

**Comment by Topbrains** [comment]
We need to be able to choose our style and select whatever width we want. For corporate development there are plenty of tools that will enforce length and formatting standards. 

How come a group that does not dare being opinionated about state management, is so opinionated about formatting?

**Comment by allanlaal** [comment]
`dartfmt` should respect `.editorconfig` (and that includes `indent_style`)

we no longer have to suffer, just because M$ Visual Studio chose 80chars and spaces and it trained a generation of devs to not know these can be changed

#customization is king and also offers #accessibility

**Comment by exaby73** [comment]
4 years later... No labels, no assignees... Looks like this is a dead issue and we'll never get this. At least make the hard limit to something reasonable like 100 chars instead of 80. We aren't in the age of square CRT monitors where the 80 char limit originated from. 100 chars serves well for 16:9 displays, with 120 to 150 for ulta-wide ones

**Comment by hbock-42** [comment]
I am afraid of the moment I will not be solo on the mobile (flutter) part at my job. How will I enforce the 100 line-length I am using?
I like your work @munificent , even got your book about programming pattern applied to game dev from before flutter existed, but on this one I have not seen any good point about preventing dart users to set line-length on a per project basis.

To sum up:
- Keep 80 the default
- Make it possible to set line-length per project (analysis_options, pubspec, who cares)
- Maybe don't advocate / advertise this feature so only people that really wants / needs it use it.
- Lower the score of package not following the 80 line-length rule on pubdev.

In the meantime, do you guys think it would be possible to make a package that once installed, when build_runner is run, modify the vscode and intellij config to change line length?

**Comment by exaby73** [comment]
@hbock-42 You can still sort of enforce it. You can make VsCode and IntelliJ project configs available in the repo so that which ever IDE your team uses, they will have both configs with the line length set to 100. On top of that, you can have a CI step that checks this using the following:

```bash
dart format --line-length 100 --set-exit-if-changed lib
```

Hope this helps you :)

**Comment by hbock-42** [comment]
> ```
>  You can make VsCode and IntelliJ project configs available in the repo so that which ever IDE your team uses, they will have both configs with the line length set to 100.
> ```

I think I found a file that might be the right candidate: `Project.xml`.
Inside this file there is a section that seems to be related to dart line length (need to make some test to validate this is not a random `120` that has nothing about the 120 line length I set in the IDE).
The file `codeStyleConfig.xml` should probably also be tracked by git because of the `PREFERRED_PROJECT_CODE_STYLE` that can be set to `Default` (which is the user IDE default code style). We want it to be `Project` (again need to check if this is really the setting I think it is referring to).

So for intellij users, this would only change IDE settings for the project, and only enforce line length (which we want - we don't want to enforce other rules).
Regarding vscode, I will have to make some research (I personally don't use it when doing flutter stuffs), I hope there is a file that contains only difference from the default user config we want to enforce, and not a file containing all vscode settings.

**Comment by crimsonvspurple** [comment]
It has been discussed already. Just commit the `Project.xml` in your git repo for IntelliJ (Dart right margin) and `settings.json` for VSCode (dart line length).

**Comment by benthillerkus** [comment]
Just saw this tracking issue for the upcoming _tall style_ https://github.com/dart-lang/dart_style/issues/1402#issue-2147897923

And some of the remaining tasks

> The FormatCommand class which drives the dart format command-line tool and accesses the file system directly should determine the language version of each file it processes. This means looking for either a pubspec or package configuration file in surrounding directories to determine the default language version.

and
> Talk to the analyzer and other IDE integration folks about updating their integration to use that new API and pass in language versions when requesting formatting.

sound a lot like the concerns raised against this configuration option.

It seems specifying formatting options in the pubspec is now viable?

**Comment by munificent** [comment]
Yes, it's definitely more viable and something I hope to do (which is why this issue is still open). However, we won't be able to reuse much if any of the code used to find the language version. We pull the language version from the `.dart_tool/package_config.yaml` file, and there is an existing package which does the work to walk the parent directories looking for that file.

For configuring the line length, I think the right place to put that is in the `analysis_options.yaml` file (or maybe the pubspec), so we'll need some different code to find that. Also, we'll have to make sure that all of the widely used tools that use dart_style as a library also correctly read that configuration and pass the proper line length in. Otherwise, you'll get different formatting results depending on whether you format within your IDE or from running `dart format ...`.

**Comment by IamRezaMousavi** [comment]
+1

**Comment by munificent** [comment]
I'm going to close this as a duplicate of #833 which is older and has more reactions on it.

**Comment by reza00farjam** [comment]
As of today's Dart update ([version 3.7.0](https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#370)), the preferred formatting page width can now be configured project-wide in the `analysis_options.yaml` file, like this:

```yml
formatter:  
  page_width: 100
```
